### ISSUE TITLE
Migrate Microsoft.Extensions packages into their own Framework (and workload)

### ISSUE THREAD

**Issue by AraHaan** [dotnet/runtime#67487]
For starters, the Microsoft.Extensions.* packages that are from this repository are great, well known, and also well used (with a majority used by tons of packages that result in a huge dependency chain of just those packages alone if you cant transitive dependencies).

Because of this I would like to suggest making them into an framework named ``Microsoft.NETCore.Extensions.App`` framework that can be used for all types of apps, This framework would replace the nuget package versions and support .NET 6 at minimum, also make them into an workload that can be installed into the .NET 7 SDK (not the .NET 6 one).

Other changes that would need to be made:
- Migrate the usage of them in dotnet/aspnetcore (and possibly other repos) to the workload version (An Sdk change would need to be made to make it know about the new workload then).
- A nuget.org change would need to be made to make it possible to obsolete the Microsoft.Extensions nuget package for that workload so they get warned to "install the workload instead".

Pros of the change:
- the libraries could benefit from crossgen2 for startup performance and various other things that might make them able to be optimized more than what the package version can offer.
- more easier to use and less chance of a dependency version conflict (the updates would flow in when you update the workload and would update for all projects) (less chance of dll hell at runtime too).
- overall the workload experience would be better than the package approach here for these well known packages that can become their own framework (because there are a lot of them that it would be more logical to make it a framework).

Cons of the change:
- Might not support as many TFMs as the package version (unless there is a way to preserve that one too).
- Would require some help to configure the workload installer generation for this in this repository, however the workload would probably need it's own SDK to insert itself into ``KnownFrameworkReference`` and define a property responsible for enabling the insertion and framework reference.

**Comment by teo-tsirpanis** [comment]
The packages in the `Microsoft.Extensions` family are plain-old general-purpose libraries. They need no special SDK support to warrant their own workload. And they also have to support .NET Framework through .NET Standard 2.0, so obsoleting the NuGet cannot happen.

**Comment by AraHaan** [comment]
Hmm I guess the framework option could work for .NET Standard based frameworks right?

Or is ``FrameworkReference`` not possible for sdk style .NET Framework projects?

**Comment by teo-tsirpanis** [comment]
I still don't understand why to make them on their own shared framework. Extensions is a library, not an app model like ASP.NET Core, Windows Desktop or MAUI. Imagine needing a shared framework for Newtonsoft.Json.

**Comment by AraHaan** [comment]
Newtonsoft is a single package, Extensions is not.

**Comment by teo-tsirpanis** [comment]
That is a miconception. Extensions is like the Windows Community Toolkit. A set of _libraries_ that could have well resided in the BCL or in a third-party library. There's absolutely nothing special about them.

**Comment by AraHaan** [comment]
While that may be the case, take a look at the ASP.NET Core framework reference, almost (if not all of them) are included in it, however they could have been instead moved to their own framework that one would have to opt-into using (with asp.net core projects opting into it by default). Because of that many people thought it was a great idea to try to manually reference the ASP.NET Core framework when using:
- WindowsDesktop
- MAUI
However it would then result in failure to run the program due to that (because they want the extensions to be used from a framework and not from a package because their application is FDD and not self-contained).

That was the main reason for filing the issue, along with the reasons I listed above like enabling crossgen2 on all the Microsoft.Extensions assemblies that are inside the reference to boost statup performance (which can also indirectly improve asp.net core / efcore as well).

**Comment by ericstj** [comment]
@eerhardt @davidfowl @maryamariyan -- thoughts on this?  One thing we could consider is adding more extensions packages to the base shared framework rather than ASP.NETCore shared framework and keep the package/framework duality.

**Comment by eerhardt** [comment]
I don't think we will get rid of the NuGet packages. We would need some pretty strong reasoning to justify removing the NuGet packages.

I agree with @ericstj above that if we think this justifies its own "shared framework", we should just move them into the Microsoft.NETCore.App shared framework. These libraries already exist in dotnet/runtime. There would be no real benefit to splitting them into their own shared framework.

But I don't think the above is strong enough justification for doing this work. If you use these libraries outside of ASP.NET, and you need fast startup, you can publish your app with `PublishReadyToRun=true`, and these assemblies will be pre-compiled. If you need to share them across multiple apps in order to save size, we should enable that scenario generically for any NuGet packages before solving it specially for Microsoft.Extensions. For example, the [`dotnet store`](https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-store) command was built to do something like this. If reusing libraries in a common location is necessary, something like that will solve it for everybody - not just Microsoft.Extensions.

**Comment by AraHaan** [comment]
What I wanted was to have them in something like a framework where they would get serviced for all of my code at once so the serviced code gets run on all of them whenever the .NET SDK updates, a lot of people could agree with me on this one.

Placing it in the base shared framework would be a good option too. While the packages might be great (they get copied to the output directory which is not always the best option for everyone). After all not everyone keeps track of package updates and that can result in security problems in their app where placing them in a shared framework would eliminate that issue as then updating the shared framework when a servicing release is issued would be a far more better (and safer) option. This is why I am reluctant to use the packages as a lot of people do not wildcard them with ``*-*`` because sometimes if they do a prerelease build might be breaking for them.

Besides they get versioned with the rest of the projects in this repo so I think having both the package option and having them in the base shared framework would be the best option (which also exists for System.Text.Json and other libraries in the base shared framework).

**Comment by eerhardt** [comment]
We can leave this issue open for a while to get feedback from others in the community if they see high value in doing this work. But without overwhelming feedback that putting the Microsoft.Extensions assemblies into the shared framework would benefit a lot of users, I don't think it is likely to get implemented.

> After all not everyone keeps track of package updates and that can result in security problems in their app where placing them in a shared framework would eliminate that issue as then updating the shared framework when a servicing release is issued would be a far more better (and safer) option.

Again, the Microsoft.Extensions libraries are not special in this regard. A security issue could come be found in any library. There are other mechanisms for finding security vulnerabilities in your dependencies. Here's one example: https://devblogs.microsoft.com/nuget/how-to-scan-nuget-packages-for-security-vulnerabilities/.

**Comment by davidfowl** [comment]
I'm open to that discussion 😃

**Comment by AraHaan** [comment]
Another option could be to do something similar to how the ``WindowsDesktop`` SDK has profiles and do an ``Extensions`` profile on the base shared framework that does not get used by default and for those who use that profile to gather the telemetry on (unless they opted out by environment variable).

**Comment by AraHaan** [comment]
What I can think of so far:

Pros:
- Projects can use the types and members from the Extensions assemblies without needing the packages.
- Projects can get less files copied to their outputs when FDD.
- Projects that normally do not maintain nuget dependencies would get free servicing updates when either Visual Studio or Windows Update patches their runtime major version (security and bug fixes that they would otherwise miss out on).
- Less disc space wasted downloading the packages as they would be part of the base shared framework instead.
- Less network usage (and faster restore) as the packages would become optional (unless they want to either roll back a breaking change or opt into the daily package builds that are not part of any daily sdk / runtime installer builds or not part of what the install scripts would install as it is too new)
- Could simplify the Dependencies in the aspnetcore repository (and others) that use the package versions (they could use the ones in the base shared framework in .NET 7+). The change would mean that the ASP.NET Core framework assemblies list would get smaller, which means in turn that it's own runtime would also be smaller (because they would be moved to the base shared framework for the ones that exist in that framework reference that are package references to the ones in this repository)  which in turn allows them to be used for any projects due to the fact they are generic enough where they can be used for any project.
- Could resolve problems that the package version might have (example could be that an extensions package might have not been patched to remove the S.R.CS.U dependency and nobody noticed it until they installed the package because the S.R.CS.U dependency no longer gets updated packages released). AKA the framework reference might catch this and will immediately get patched to then not include the S.R.CS.U dependency at all.

Cons:
- Would not handle fresh builds in the nuget feed which are not yet part of the daily feed builds that the install script/installers would install (forcing package to be used until then for some).
- The Extensions packages that exist in the aspnetcore repository would not be possible to be included in the base shared framework (due to the fact it was made specifically for ASP.NET Core)

**Comment by jkotas** [comment]
> Less disc space wasted downloading the packages as they would be part of the base shared framework instead.

It is not true if your project depends on netcoreapp shared framework only and does not use any of these packages. We do publish packages with netcoreapp shared framework only (https://dotnet.microsoft.com/en-us/download/dotnet/6.0).

**Comment by AraHaan** [comment]
Yep, that was another reason why I opened this issue as well, to discuss a way to and permanently resolve the problem via roll forward logic that already exists in the way the runtime works.

**Comment by ericstj** [comment]
Moving discussion from dotnet#4669 to here.

@agocke mentions
> I don’t understand: what is the plan for users of self-contained apps with no ASP reference? How can they avoid a regression?

Any feature addition to the framework grows the size of the framework for people not using that feature.  Where does the line get drawn?  For folks who really care about size we have the linker.  Everything moved here would be linked out if not used.  We don't have promises around the size of self-contained apps that don't use the linker between major versions of a release.  They see growth (regression) with pretty much every release.

## Size of self-contained hello-world across different framework versions
| Target Framework | linux-x64 | win-x64 |
|---|---|---|
| netcoreapp2.1 | 70.7 MB (184 files) | 65.34 MB (217 files) |
| netcoreapp3.1 | 74.89 MB (191 files) | 65.75 MB (225 files) |
| net5.0 | 72.11 MB (186 files) | 65.11 MB (225 files) |
| net6.0 | 69.04 MB (184 files) | 68.72 MB (227 files) |
| net7.0 | 71.49 MB (187 files) | 69.61 MB (187 files) |
| net8.0 | 70.56 MB (187 files) | 70.56 MB (187 files) |
| net9.0 | 75.97 MB (189 files) | 74.38 MB (189 files) |
| net10.0 | 78.76 MB (192 files) | 76.99 MB (192 files)|

Reducing the size/technology-count/build-reference-count growth was one of the reasons I mentioned above that I'm dialing this back since I don't think all the side-effects outweighed the benefits.  I've been more concerned with the build perf and technology prolifieration since folks can mitigate runtime pack size growth with linker -- but you are right to raise it as it does impact some.  The size of the a few of the options being considered is as follows - measuring cross-gened size for RIDs, app-local files for IL, counts are runtime files but are mostly of the same order as reference files.

## Size of proposed assembly set moves
| Group | linux-x64 | win-x64 | IL size
|---|---|---|---|
Hosting-closure (30 references) | 2.3 MB (31 files) |  3.7 MB (32 files) | 2.3 MB (33 files)
Abstractions (9 references) | 698 KB (9 files) | 793 KB (9 files) | 412 KB (9 files)
DI/Logging/Options Abstractions (4 references) | 468 KB (4 files) | 509 KB (4 files) | 239 KB (4 files)
DI/Logging Abstractions (2 references) |  269 KB (2 files) |  289 KB (2 files) |  131 KB (2 files) 

My early data gathering from library dependencies / usage pointed to _DI/Logging/Options Abstractions_ , however we saw another grouping of download usage at the hosting closure which is why we considered that larger set as the first try.  I've decided against the larger set as mentioned here, but still considering which of the remaining three.

**Comment by rosebyte** [comment]
Closing the issue as everything seems completed and there isn't any visible actionable work left.
